pysqream_blue/cursor.py

```diff
@@ -157,66 +165,77 @@ def _compile(self, query):
         self.stmt_id, self.columns_metadata, self.query_type = response.context_id, response.columns, response.query_type
         return response
 
-    def _request_execute(self):
-        try:
-            execute_response: qh_messages.ExecuteResponse = self._execute()
-
-            if execute_response.HasField('error'):
-                self.logs.log_and_raise(OperationalError,
-                              f'Query id: {self.stmt_id}. Error while attempting to execute the query.\n{execute_response.error}')
-
-            self.logs.message(f'Query id: {self.stmt_id}. The query was send to execute successfully.', self.logs.info)
+    def _request_execute_inner(self):
+        execute_response: qh_messages.ExecuteResponse = self._execute()
+        if execute_response.HasField('error'):
+            self.logs.log_and_raise(OperationalError,
+                          f'Query id: {self.stmt_id}. Error while attempting to execute the query.\n{execute_response.error}')
+        self.logs.message(f'Query id: {self.stmt_id}. The query was send to execute successfully.', self.logs.info)
 
-        except grpc.RpcError as rpc_error:
-            self.logs.log_and_raise(ProgrammingError,
-                          f'Query id: {self.stmt_id}. Error from grpc while attempting to execute the query.\n{rpc_error}')
-            self.channel.close()
+    def _request_execute(self):
+        self._retry_function_on_stream_error(self._request_execute_inner, "execute the query")
 
     def _execute(self):
         return self.client.Execute(
             qh_messages.ExecuteRequest(context_id=self.stmt_id),
             credentials=self.call_credentialds if self.use_ssl else None)
 
-    def _request_status(self):
+    def _retry_function_on_stream_error(self, func, function_action_name, *args):
+        number_of_retries = 3
+        for i in range(number_of_retries):
+            try:
+                if len(args) == 0:
+                    return func()
+                elif len(args) == 1:
+                    return func(args[0])
+                else:
+                    self.logs.log_and_raise(ProgrammingError, "An illegal number of args were passed to _retry_function_on_stream_error", self.logs.error)
+            except grpc.RpcError as rpc_error:
+                if "Stream removed" not in rpc_error.details() or i == number_of_retries-1:
+                    break
+                self.logs.message("GRPC Error encountered - stream removed. Retrying.", self.logs.error)
+                self._restart_channel_and_client()
+                continue
+        self.channel.close()
+        self.logs.log_and_raise(ProgrammingError,
+                              f'Query id: {self.stmt_id}. Error from grpc while attempting to {function_action_name}.\n{rpc_error}')
 
-        self.logs.message(f'Query id: {self.stmt_id}. Wait for execution.', self.logs.info)
+    def _request_status_inner(self):
```

Comment:
We can remove the while True and create a better sleep decorator - 

```
def wait():
     def decorator(func):
         def wrapper(*args, **kwargs):
            self. _smart_sleep()
             return func(*args, **kwargs)
         return wrapper
     return decorator
```

You need before calling `_request_status_inner` to init `self.try_status_num = 0`
Or that you sent into the wait function argument and then use inside `self. _smart_sleep()`.

for using wait - 
```
@wait()
_request_status_inner(...)
```